daprdocs/content/en/developing-applications/building-blocks/service-invocation/howto-invoke-services-grpc.md

```diff
@@ -0,0 +1,238 @@
+)
+
+const (
+	port = ":50051"
+)
+
+// server is used to implement helloworld.GreeterServer.
+type server struct {
+	pb.UnimplementedGreeterServer
+}
+
+// SayHello implements helloworld.GreeterServer
+func (s *server) SayHello(ctx context.Context, in *pb.HelloRequest) (*pb.HelloReply, error) {
+	log.Printf("Received: %v", in.GetName())
+	return &pb.HelloReply{Message: "Hello " + in.GetName()}, nil
+}
+
+func main() {
+	lis, err := net.Listen("tcp", port)
+	if err != nil {
+		log.Fatalf("failed to listen: %v", err)
+	}
+	s := grpc.NewServer()
+	pb.RegisterGreeterServer(s, &server{})
+	log.Printf("server listening at %v", lis.Addr())
+	if err := s.Serve(lis); err != nil {
+		log.Fatalf("failed to serve: %v", err)
+	}
+}
+```
+
+This Go app implements the Greeter proto service and exposes a `SayHello` method.
+
+### Run the gRPC server using the Dapr CLI
+
+Since gRPC proxying is an opt-in feature, first enable it using a configuration file:
+
+```yaml
+apiVersion: dapr.io/v1alpha1
+kind: Configuration
+metadata:
+  name: serverconfig
+spec:
+  tracing:
+    samplingRate: "1"
+    zipkin:
+      endpointAddress: http://localhost:9411/api/v2/spans
+  features:
+    - name: proxy.grpc
+      enabled: true
+```
+
+Run the sidecar and the Go server:
+
+```bash
+dapr run --app-id server --app-protocol grpc --app-port 50051 --config config.yaml -- go run main.go
+```
+
+Using the Dapr CLI, we're assigning a unique id to the app, `server`, using the `--app-id` flag.
+
+## Step 2: Invoke the service
+
+The following example shows you how to discover the Greeter service using Dapr from a gRPC client.
+Notice that instead of invoking the target service directly at port `50051`, the client is invoking its local Dapr sidecar over port `50007` which then takes care of service discovery, telemetry, mTLS and other concerns.
+
+```go
+package main
+
+import (
+	"context"
+	"log"
+	"time"
+
+	"google.golang.org/grpc"
+	pb "google.golang.org/grpc/examples/helloworld/helloworld"
+	"google.golang.org/grpc/metadata"
+)
+
+const (
+	address = "localhost:50007"
+)
+
+func main() {
+	// Set up a connection to the server.
+	conn, err := grpc.Dial(address, grpc.WithInsecure(), grpc.WithBlock())
+	if err != nil {
+		log.Fatalf("did not connect: %v", err)
+	}
+	defer conn.Close()
+	c := pb.NewGreeterClient(conn)
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*2)
+	defer cancel()
+
+	ctx = metadata.AppendToOutgoingContext(ctx, "dapr-app-id", "server")
+	r, err := c.SayHello(ctx, &pb.HelloRequest{Name: "Darth Tyrannus"})
+	if err != nil {
+		log.Fatalf("could not greet: %v", err)
+	}
+
+	log.Printf("Greeting: %s", r.GetMessage())
+}
+```
+
+The following line tells Dapr to discover and invoke an app named `server`:
+
+```go
+ctx = metadata.AppendToOutgoingContext(ctx, "dapr-app-id", "server")
+```
+
+### Run the client using the Dapr CLI
+
+Since gRPC proxying is an opt-in feature, first enable it using a configuration file:
+
```

Comment:
"Since gRPC proxying is a preview feature in the v1.3 release, you need to opt-in using a configuration file. See https://docs.dapr.io/operations/configuration/preview-features/ for more information.

Reply:
Done. Changed to "currently in preview" to not tie it to any particular release.

Reply:
Remember to update this in the future release.